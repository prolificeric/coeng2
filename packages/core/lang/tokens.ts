export type TokenType =
  | 'ATOM'
  | 'BRANCH_SEPARATOR'
  | 'PART_SEPARATOR'
  | 'L_PAREN'
  | 'R_PAREN'
  | 'L_CURLY'
  | 'R_CURLY'
  | 'L_SQUARE'
  | 'R_SQUARE'
  | 'HEAD_REF'
  | 'PREV_SEQ_REF'
  | 'SORTED_SET_INIT';

export type Token = {
  type: TokenType;
  value: string;
  loc: {
    start: Cursor;
    end: Cursor;
  };
};

export type Cursor = {
  offset: number;
  line: number;
  column: number;
};

export type TokenParser = {
  type: TokenType;
  parse: (input: string) => string | null;
};

export const RegExpParser = (
  regexp: RegExp,
  mod?: (match: string[]) => string,
): TokenParser['parse'] => {
  return (input: string) => {
    const match = regexp.exec(input) ?? null;
    return match ? (mod ? mod(match) : match[0]) : null;
  };
};

export const defaultTokenParsers: TokenParser[] = [
  {
    type: 'BRANCH_SEPARATOR',
    parse: RegExpParser(/^[,\n][, \t\n]*/),
  },
  {
    type: 'PART_SEPARATOR',
    parse: RegExpParser(/^[\t ]*/),
  },
  {
    type: 'L_PAREN',
    parse: RegExpParser(/^\([\n\t, ]*/),
  },
  {
    type: 'R_PAREN',
    parse: RegExpParser(/^[\n\t, ]*\)/),
  },
  {
    type: 'L_CURLY',
    parse: RegExpParser(/^\{[\n\t, ]*/),
  },
  {
    type: 'R_CURLY',
    parse: RegExpParser(/^[\n\t, ]*\}/),
  },
  {
    type: 'L_SQUARE',
    parse: RegExpParser(/^\[[\n\t, ]*/),
  },
  {
    type: 'R_SQUARE',
    parse: RegExpParser(/^[\n\t, ]*\]/),
  },
  {
    // [: z b a c] -> [:a b c z]
    type: 'SORTED_SET_INIT',
    parse: RegExpParser(/^:/),
  },
  {
    // john { knows mary (knows &) } -> john knows mary, mary knows john
    type: 'HEAD_REF',
    parse: RegExpParser(/^&/),
  },
  {
    // john knows {max, mary ([...] since 2000)} -> john knows max, [john knows mary] since 2000
    type: 'PREV_SEQ_REF',
    parse: RegExpParser(/^\.{3}(?!\.)/),
  },
  {
    // x..y
    type: 'PREV_SEQ_REF',
    parse: RegExpParser(/^[0-9]+\.{2}[0-9]+/),
  },
  {
    // ..y
    type: 'PREV_SEQ_REF',
    parse: RegExpParser(/^[0-9]+\.{2}(?!\.)/),
  },
  {
    // x..
    type: 'PREV_SEQ_REF',
    parse: RegExpParser(/^\.{2}[0-9]+/),
  },
  {
    // <<arbitrary string goes here>>
    type: 'ATOM',
    parse: RegExpParser(/^<<(.|\n)*?>>/),
  },
  {
    type: 'ATOM',
    parse: RegExpParser(/^[^ ,\n\(\)\{\}\[\]]+/),
  },
];

export const tokenize = (
  input: string,
  tokenParsers: TokenParser[] = defaultTokenParsers,
): Token[] => {
  const trimmedLeft = input.trimStart();

  const trimmedRight = trimmedLeft.trimEnd();

  if (!trimmedRight) {
    return [];
  }

  const trimmedLines = input.slice(0, trimmedLeft.length).split('\n');

  let start: Cursor = {
    offset: input.length - trimmedLeft.length,
    line: trimmedLines.length,
    column: trimmedLines.at(-1)!.length + 1,
  };

  const tokens: Token[] = [];

  input = trimmedRight;

  inputLoop: while (input) {
    tokenParserLoop: for (const { type, parse } of tokenParsers) {
      const value = parse(input);

      if (!value) {
        continue tokenParserLoop;
      }

      const lines = value.split('\n');

      const end: Cursor = {
        offset: start.offset + value.length,
        line: start.line + lines.length - 1,
        column:
          lines.length > 1
            ? lines[lines.length - 1].length + 1
            : start.column + value.length,
      };

      const token: Token = {
        type,
        value,
        loc: { start, end },
      };

      tokens.push(token);

      input = input.slice(value.length);
      start = end;

      continue inputLoop;
    }

    throw new Error(
      `Unexpected token at line ${start.line}, column ${
        start.column
      }: ${JSON.stringify(input.slice(0, 10))}`,
    );
  }

  return tokens;
};
