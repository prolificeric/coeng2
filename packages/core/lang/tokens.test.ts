import { describe, test, expect } from 'bun:test';
import { RegExpParser, tokenize } from './tokens';

describe('parse', () => {
  describe('RegExpParser', () => {
    test('should return null if no match', () => {
      const parser = RegExpParser(/^foo/);
      expect(parser('bar')).toBeNull();
    });

    test('should return the matched string', () => {
      const parser = RegExpParser(/^foo/);
      expect(parser('foobar')).toBe('foo');
    });
  });

  describe('tokenize', () => {
    test('should return an empty array if the input is empty', () => {
      expect(tokenize('')).toEqual([]);
    });

    test('should return an empty array if the input is whitespace', () => {
      expect(tokenize(' ')).toEqual([]);
      expect(tokenize('\n')).toEqual([]);
      expect(tokenize('\t')).toEqual([]);
      expect(tokenize('\n\n')).toEqual([]);
      expect(tokenize('\n\n\n\n\n\n\n\n')).toEqual([]);
    });

    test('works on complex input', () => {
      const tokens = tokenize(`
        eric-weber {
          knows {
            typescript
            javascript
            [... excel (spreadsheet-app)] since 1999
          } (programming-language)
    
          named     <<Eric Weber>>
    
          described-as <<a person
          
          
          who writes code>>

          john knows &
        }
    
        [eric knows javascript] since 1999

        [: Hi I'm a sorted set!]
      `);

      expect(tokens).toMatchSnapshot();
    });
  });
});
