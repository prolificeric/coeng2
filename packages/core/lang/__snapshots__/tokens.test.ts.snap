// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse tokenize works on complex input 1`] = `
[
  {
    "loc": {
      "end": {
        "column": 41,
        "line": 21,
        "offset": 19,
      },
      "start": {
        "column": 31,
        "line": 21,
        "offset": 9,
      },
    },
    "type": "ATOM",
    "value": "eric-weber",
  },
  {
    "loc": {
      "end": {
        "column": 42,
        "line": 21,
        "offset": 20,
      },
      "start": {
        "column": 41,
        "line": 21,
        "offset": 19,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 11,
        "line": 22,
        "offset": 32,
      },
      "start": {
        "column": 42,
        "line": 21,
        "offset": 20,
      },
    },
    "type": "L_CURLY",
    "value": 
"{
          "
,
  },
  {
    "loc": {
      "end": {
        "column": 16,
        "line": 22,
        "offset": 37,
      },
      "start": {
        "column": 11,
        "line": 22,
        "offset": 32,
      },
    },
    "type": "ATOM",
    "value": "knows",
  },
  {
    "loc": {
      "end": {
        "column": 17,
        "line": 22,
        "offset": 38,
      },
      "start": {
        "column": 16,
        "line": 22,
        "offset": 37,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 13,
        "line": 23,
        "offset": 52,
      },
      "start": {
        "column": 17,
        "line": 22,
        "offset": 38,
      },
    },
    "type": "L_CURLY",
    "value": 
"{
            "
,
  },
  {
    "loc": {
      "end": {
        "column": 23,
        "line": 23,
        "offset": 62,
      },
      "start": {
        "column": 13,
        "line": 23,
        "offset": 52,
      },
    },
    "type": "ATOM",
    "value": "typescript",
  },
  {
    "loc": {
      "end": {
        "column": 13,
        "line": 24,
        "offset": 75,
      },
      "start": {
        "column": 23,
        "line": 23,
        "offset": 62,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
            "
,
  },
  {
    "loc": {
      "end": {
        "column": 23,
        "line": 24,
        "offset": 85,
      },
      "start": {
        "column": 13,
        "line": 24,
        "offset": 75,
      },
    },
    "type": "ATOM",
    "value": "javascript",
  },
  {
    "loc": {
      "end": {
        "column": 13,
        "line": 25,
        "offset": 98,
      },
      "start": {
        "column": 23,
        "line": 24,
        "offset": 85,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
            "
,
  },
  {
    "loc": {
      "end": {
        "column": 14,
        "line": 25,
        "offset": 99,
      },
      "start": {
        "column": 13,
        "line": 25,
        "offset": 98,
      },
    },
    "type": "L_SQUARE",
    "value": "[",
  },
  {
    "loc": {
      "end": {
        "column": 17,
        "line": 25,
        "offset": 102,
      },
      "start": {
        "column": 14,
        "line": 25,
        "offset": 99,
      },
    },
    "type": "PREV_SEQ_REF",
    "value": "...",
  },
  {
    "loc": {
      "end": {
        "column": 18,
        "line": 25,
        "offset": 103,
      },
      "start": {
        "column": 17,
        "line": 25,
        "offset": 102,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 23,
        "line": 25,
        "offset": 108,
      },
      "start": {
        "column": 18,
        "line": 25,
        "offset": 103,
      },
    },
    "type": "ATOM",
    "value": "excel",
  },
  {
    "loc": {
      "end": {
        "column": 24,
        "line": 25,
        "offset": 109,
      },
      "start": {
        "column": 23,
        "line": 25,
        "offset": 108,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 25,
        "line": 25,
        "offset": 110,
      },
      "start": {
        "column": 24,
        "line": 25,
        "offset": 109,
      },
    },
    "type": "L_PAREN",
    "value": "(",
  },
  {
    "loc": {
      "end": {
        "column": 40,
        "line": 25,
        "offset": 125,
      },
      "start": {
        "column": 25,
        "line": 25,
        "offset": 110,
      },
    },
    "type": "ATOM",
    "value": "spreadsheet-app",
  },
  {
    "loc": {
      "end": {
        "column": 41,
        "line": 25,
        "offset": 126,
      },
      "start": {
        "column": 40,
        "line": 25,
        "offset": 125,
      },
    },
    "type": "R_PAREN",
    "value": ")",
  },
  {
    "loc": {
      "end": {
        "column": 42,
        "line": 25,
        "offset": 127,
      },
      "start": {
        "column": 41,
        "line": 25,
        "offset": 126,
      },
    },
    "type": "R_SQUARE",
    "value": "]",
  },
  {
    "loc": {
      "end": {
        "column": 43,
        "line": 25,
        "offset": 128,
      },
      "start": {
        "column": 42,
        "line": 25,
        "offset": 127,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 48,
        "line": 25,
        "offset": 133,
      },
      "start": {
        "column": 43,
        "line": 25,
        "offset": 128,
      },
    },
    "type": "ATOM",
    "value": "since",
  },
  {
    "loc": {
      "end": {
        "column": 49,
        "line": 25,
        "offset": 134,
      },
      "start": {
        "column": 48,
        "line": 25,
        "offset": 133,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 53,
        "line": 25,
        "offset": 138,
      },
      "start": {
        "column": 49,
        "line": 25,
        "offset": 134,
      },
    },
    "type": "ATOM",
    "value": "1999",
  },
  {
    "loc": {
      "end": {
        "column": 11,
        "line": 26,
        "offset": 149,
      },
      "start": {
        "column": 53,
        "line": 25,
        "offset": 138,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
          "
,
  },
  {
    "loc": {
      "end": {
        "column": 12,
        "line": 26,
        "offset": 150,
      },
      "start": {
        "column": 11,
        "line": 26,
        "offset": 149,
      },
    },
    "type": "R_CURLY",
    "value": "}",
  },
  {
    "loc": {
      "end": {
        "column": 13,
        "line": 26,
        "offset": 151,
      },
      "start": {
        "column": 12,
        "line": 26,
        "offset": 150,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 14,
        "line": 26,
        "offset": 152,
      },
      "start": {
        "column": 13,
        "line": 26,
        "offset": 151,
      },
    },
    "type": "L_PAREN",
    "value": "(",
  },
  {
    "loc": {
      "end": {
        "column": 34,
        "line": 26,
        "offset": 172,
      },
      "start": {
        "column": 14,
        "line": 26,
        "offset": 152,
      },
    },
    "type": "ATOM",
    "value": "programming-language",
  },
  {
    "loc": {
      "end": {
        "column": 35,
        "line": 26,
        "offset": 173,
      },
      "start": {
        "column": 34,
        "line": 26,
        "offset": 172,
      },
    },
    "type": "R_PAREN",
    "value": ")",
  },
  {
    "loc": {
      "end": {
        "column": 11,
        "line": 28,
        "offset": 189,
      },
      "start": {
        "column": 35,
        "line": 26,
        "offset": 173,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
    
          "
,
  },
  {
    "loc": {
      "end": {
        "column": 16,
        "line": 28,
        "offset": 194,
      },
      "start": {
        "column": 11,
        "line": 28,
        "offset": 189,
      },
    },
    "type": "ATOM",
    "value": "named",
  },
  {
    "loc": {
      "end": {
        "column": 21,
        "line": 28,
        "offset": 199,
      },
      "start": {
        "column": 16,
        "line": 28,
        "offset": 194,
      },
    },
    "type": "PART_SEPARATOR",
    "value": "     ",
  },
  {
    "loc": {
      "end": {
        "column": 35,
        "line": 28,
        "offset": 213,
      },
      "start": {
        "column": 21,
        "line": 28,
        "offset": 199,
      },
    },
    "type": "ATOM",
    "value": "<<Eric Weber>>",
  },
  {
    "loc": {
      "end": {
        "column": 11,
        "line": 30,
        "offset": 229,
      },
      "start": {
        "column": 35,
        "line": 28,
        "offset": 213,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
    
          "
,
  },
  {
    "loc": {
      "end": {
        "column": 23,
        "line": 30,
        "offset": 241,
      },
      "start": {
        "column": 11,
        "line": 30,
        "offset": 229,
      },
    },
    "type": "ATOM",
    "value": "described-as",
  },
  {
    "loc": {
      "end": {
        "column": 24,
        "line": 30,
        "offset": 242,
      },
      "start": {
        "column": 23,
        "line": 30,
        "offset": 241,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 28,
        "line": 33,
        "offset": 302,
      },
      "start": {
        "column": 24,
        "line": 30,
        "offset": 242,
      },
    },
    "type": "ATOM",
    "value": 
"<<a person
          
          
          who writes code>>"
,
  },
  {
    "loc": {
      "end": {
        "column": 11,
        "line": 35,
        "offset": 314,
      },
      "start": {
        "column": 28,
        "line": 33,
        "offset": 302,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"

          "
,
  },
  {
    "loc": {
      "end": {
        "column": 15,
        "line": 35,
        "offset": 318,
      },
      "start": {
        "column": 11,
        "line": 35,
        "offset": 314,
      },
    },
    "type": "ATOM",
    "value": "john",
  },
  {
    "loc": {
      "end": {
        "column": 16,
        "line": 35,
        "offset": 319,
      },
      "start": {
        "column": 15,
        "line": 35,
        "offset": 318,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 21,
        "line": 35,
        "offset": 324,
      },
      "start": {
        "column": 16,
        "line": 35,
        "offset": 319,
      },
    },
    "type": "ATOM",
    "value": "knows",
  },
  {
    "loc": {
      "end": {
        "column": 22,
        "line": 35,
        "offset": 325,
      },
      "start": {
        "column": 21,
        "line": 35,
        "offset": 324,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 23,
        "line": 35,
        "offset": 326,
      },
      "start": {
        "column": 22,
        "line": 35,
        "offset": 325,
      },
    },
    "type": "HEAD_REF",
    "value": "&",
  },
  {
    "loc": {
      "end": {
        "column": 9,
        "line": 36,
        "offset": 335,
      },
      "start": {
        "column": 23,
        "line": 35,
        "offset": 326,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
        "
,
  },
  {
    "loc": {
      "end": {
        "column": 10,
        "line": 36,
        "offset": 336,
      },
      "start": {
        "column": 9,
        "line": 36,
        "offset": 335,
      },
    },
    "type": "R_CURLY",
    "value": "}",
  },
  {
    "loc": {
      "end": {
        "column": 9,
        "line": 38,
        "offset": 350,
      },
      "start": {
        "column": 10,
        "line": 36,
        "offset": 336,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"
    
        "
,
  },
  {
    "loc": {
      "end": {
        "column": 10,
        "line": 38,
        "offset": 351,
      },
      "start": {
        "column": 9,
        "line": 38,
        "offset": 350,
      },
    },
    "type": "L_SQUARE",
    "value": "[",
  },
  {
    "loc": {
      "end": {
        "column": 14,
        "line": 38,
        "offset": 355,
      },
      "start": {
        "column": 10,
        "line": 38,
        "offset": 351,
      },
    },
    "type": "ATOM",
    "value": "eric",
  },
  {
    "loc": {
      "end": {
        "column": 15,
        "line": 38,
        "offset": 356,
      },
      "start": {
        "column": 14,
        "line": 38,
        "offset": 355,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 20,
        "line": 38,
        "offset": 361,
      },
      "start": {
        "column": 15,
        "line": 38,
        "offset": 356,
      },
    },
    "type": "ATOM",
    "value": "knows",
  },
  {
    "loc": {
      "end": {
        "column": 21,
        "line": 38,
        "offset": 362,
      },
      "start": {
        "column": 20,
        "line": 38,
        "offset": 361,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 31,
        "line": 38,
        "offset": 372,
      },
      "start": {
        "column": 21,
        "line": 38,
        "offset": 362,
      },
    },
    "type": "ATOM",
    "value": "javascript",
  },
  {
    "loc": {
      "end": {
        "column": 32,
        "line": 38,
        "offset": 373,
      },
      "start": {
        "column": 31,
        "line": 38,
        "offset": 372,
      },
    },
    "type": "R_SQUARE",
    "value": "]",
  },
  {
    "loc": {
      "end": {
        "column": 33,
        "line": 38,
        "offset": 374,
      },
      "start": {
        "column": 32,
        "line": 38,
        "offset": 373,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 38,
        "line": 38,
        "offset": 379,
      },
      "start": {
        "column": 33,
        "line": 38,
        "offset": 374,
      },
    },
    "type": "ATOM",
    "value": "since",
  },
  {
    "loc": {
      "end": {
        "column": 39,
        "line": 38,
        "offset": 380,
      },
      "start": {
        "column": 38,
        "line": 38,
        "offset": 379,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 43,
        "line": 38,
        "offset": 384,
      },
      "start": {
        "column": 39,
        "line": 38,
        "offset": 380,
      },
    },
    "type": "ATOM",
    "value": "1999",
  },
  {
    "loc": {
      "end": {
        "column": 9,
        "line": 40,
        "offset": 394,
      },
      "start": {
        "column": 43,
        "line": 38,
        "offset": 384,
      },
    },
    "type": "BRANCH_SEPARATOR",
    "value": 
"

        "
,
  },
  {
    "loc": {
      "end": {
        "column": 10,
        "line": 40,
        "offset": 395,
      },
      "start": {
        "column": 9,
        "line": 40,
        "offset": 394,
      },
    },
    "type": "L_SQUARE",
    "value": "[",
  },
  {
    "loc": {
      "end": {
        "column": 11,
        "line": 40,
        "offset": 396,
      },
      "start": {
        "column": 10,
        "line": 40,
        "offset": 395,
      },
    },
    "type": "SORTED_SET_INIT",
    "value": ":",
  },
  {
    "loc": {
      "end": {
        "column": 12,
        "line": 40,
        "offset": 397,
      },
      "start": {
        "column": 11,
        "line": 40,
        "offset": 396,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 14,
        "line": 40,
        "offset": 399,
      },
      "start": {
        "column": 12,
        "line": 40,
        "offset": 397,
      },
    },
    "type": "ATOM",
    "value": "Hi",
  },
  {
    "loc": {
      "end": {
        "column": 15,
        "line": 40,
        "offset": 400,
      },
      "start": {
        "column": 14,
        "line": 40,
        "offset": 399,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 18,
        "line": 40,
        "offset": 403,
      },
      "start": {
        "column": 15,
        "line": 40,
        "offset": 400,
      },
    },
    "type": "ATOM",
    "value": "I'm",
  },
  {
    "loc": {
      "end": {
        "column": 19,
        "line": 40,
        "offset": 404,
      },
      "start": {
        "column": 18,
        "line": 40,
        "offset": 403,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 20,
        "line": 40,
        "offset": 405,
      },
      "start": {
        "column": 19,
        "line": 40,
        "offset": 404,
      },
    },
    "type": "ATOM",
    "value": "a",
  },
  {
    "loc": {
      "end": {
        "column": 21,
        "line": 40,
        "offset": 406,
      },
      "start": {
        "column": 20,
        "line": 40,
        "offset": 405,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 27,
        "line": 40,
        "offset": 412,
      },
      "start": {
        "column": 21,
        "line": 40,
        "offset": 406,
      },
    },
    "type": "ATOM",
    "value": "sorted",
  },
  {
    "loc": {
      "end": {
        "column": 28,
        "line": 40,
        "offset": 413,
      },
      "start": {
        "column": 27,
        "line": 40,
        "offset": 412,
      },
    },
    "type": "PART_SEPARATOR",
    "value": " ",
  },
  {
    "loc": {
      "end": {
        "column": 32,
        "line": 40,
        "offset": 417,
      },
      "start": {
        "column": 28,
        "line": 40,
        "offset": 413,
      },
    },
    "type": "ATOM",
    "value": "set!",
  },
  {
    "loc": {
      "end": {
        "column": 33,
        "line": 40,
        "offset": 418,
      },
      "start": {
        "column": 32,
        "line": 40,
        "offset": 417,
      },
    },
    "type": "R_SQUARE",
    "value": "]",
  },
]
`;
